swift/apple/FirezoneKit/Sources/FirezoneKit/Models/AuthClient.swift

```diff
@@ -0,0 +1,102 @@
+//
+//  AuthClient.swift
+//  (c) 2024 Firezone, Inc.
+//  LICENSE: Apache-2.0
+//
+
+import AuthenticationServices
+import Foundation
+
+enum AuthClientError: Error {
+  case invalidCallbackURL
+  case invalidStateReturnedInCallback(expected: String, got: String)
+  case authResponseError(Error)
+  case sessionFailure(Error)
+  case randomNumberGenerationFailure(errorStatus: Int32)
+}
+
+struct AuthClient {
+  private var authURL: URL
+  private var state: String
+  private var nonce: String
+
+  init?(authURL: URL) throws {
+    self.authURL = authURL
+    state = try Self.createRandomHexString(byteCount: 32)
+    nonce = try Self.createRandomHexString(byteCount: 32)
+  }
+
+  // Builds a full URL to send to the portal
+  func build() throws -> URL {
+    return authURL
+      .appendingQueryItem(URLQueryItem(name: "state", value: state))
+      .appendingQueryItem(URLQueryItem(name: "nonce", value: nonce))
+      .appendingQueryItem(URLQueryItem(name: "as", value: "client"))
+  }
+
+  func response(url: URL?) throws -> AuthResponse {
+    guard let url = url,
+          let returnedState = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems?.first(where: { $0.name == "state" })?.value,
+          areStringsEqualConstantTime(state, returnedState),
+          let fragment = URLComponents(url: url, resolvingAgainstBaseURL: false)?
+          .queryItems?
+          .first(where: { $0.name == "fragment" })?
+          .value,
+          let actorName = URLComponents(url: url, resolvingAgainstBaseURL: false)?
+          .queryItems?
+          .first(where: { $0.name == "actor_name" })?
+          .value?
+          .removingPercentEncoding?
+          .replacingOccurrences(of: "+", with: " ")
```

Comment:
I don't have this replace in Tauri. I thought the plus sign was part of the percent encoding or something. (though that should be `%20`, right?) This unit test could be used to see if it's really needed: https://github.com/firezone/firezone/blob/e8dc190879df0cca2a5f3e780895806d2a99a05b/rust/gui-client/src-tauri/src/client/deep_link.rs#L89

Reply:
Yeah this is needed because the `+` is percent encoded when we get it (`%2B`), so the line above removes that, and this removes the `+`

Reply:
Then it shouldn't be working in Tauri lol 😨

Reply:
`removePercentEncoding` doesn't remove the `+` either